File: apps/api-rs/src/document/cleaning.rs
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::{BTreeMap, BTreeSet, HashSet};
use unicode_normalization::UnicodeNormalization;

use super::{FileType, ParsedBlock};

pub const CLEANER_VERSION: &str = "documind-cleaner@0.2.0";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CleanedBlock {
    pub block: ParsedBlock,
    pub cleaned_text: String,
    pub is_removed: bool,
    pub remove_reason: Option<String>,
    pub cleaning_ops: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CleanStats {
    pub input_blocks: usize,
    pub output_blocks: usize,
    pub removed_blocks: usize,
    pub ops_top: Vec<String>,
}

pub fn clean_blocks(
    file_type: FileType,
    blocks: &[ParsedBlock],
) -> (Vec<CleanedBlock>, CleanStats) {
    let mut cleaned = Vec::with_capacity(blocks.len());
    let mut op_counts = BTreeMap::<String, usize>::new();
    let repeated_pdf_noise = repeated_pdf_noise(file_type, blocks);

    for block in blocks {
        let (mut text, mut ops) = common_clean(&block.text);
        format_specific_clean(file_type, block, &mut text, &mut ops);

        let (is_removed, remove_reason) =
            removal_reason(file_type, block, &text, &repeated_pdf_noise);
        for op in &ops {
            *op_counts.entry(op.clone()).or_insert(0) += 1;
        }

        cleaned.push(CleanedBlock {
            block: block.clone(),
            cleaned_text: text,
            is_removed,
            remove_reason,
            cleaning_ops: ops,
        });
    }

    let removed_blocks = cleaned.iter().filter(|b| b.is_removed).count();
    let output_blocks = cleaned.len().saturating_sub(removed_blocks);
    let mut ops_top = op_counts.into_iter().collect::<Vec<_>>();
    ops_top.sort_by(|a, b| b.1.cmp(&a.1).then_with(|| a.0.cmp(&b.0)));

    (
        cleaned,
        CleanStats {
            input_blocks: blocks.len(),
            output_blocks,
            removed_blocks,
            ops_top: ops_top.into_iter().take(8).map(|(op, _)| op).collect(),
        },
    )
}

pub fn cleaned_block_metadata(block: &CleanedBlock) -> Value {
    json!({
        "cleaner_version": CLEANER_VERSION,
        "cleaning_ops": block.cleaning_ops,
        "is_removed": block.is_removed,
        "remove_reason": block.remove_reason,
        "source_ref": block.block.source_ref,
        "source_metadata": block.block.metadata,
    })
}

fn common_clean(input: &str) -> (String, Vec<String>) {
    let mut ops = Vec::new();
    let mut text = input.to_string();

    if text.starts_with('\u{feff}') {
        text = text.trim_start_matches('\u{feff}').to_string();
        ops.push("remove_bom".to_string());
    }

    if text.contains("\r\n") || text.contains('\r') {
        text = text.replace("\r\n", "\n").replace('\r', "\n");
        ops.push("normalize_line_endings".to_string());
    }

    let before = text.clone();
    text = text
        .chars()
        .filter_map(|c| match c {
            '\u{00a0}' | '\u{3000}' => Some(' '),
            '\u{200b}' | '\u{200c}' | '\u{200d}' | '\u{2060}' | '\u{fffc}' => None,
            '\t' => Some(' '),
            '\n' => Some('\n'),
            c if c.is_control() => None,
            c => Some(c),
        })
        .collect::<String>();
    if text != before {
        ops.push("normalize_space".to_string());
        ops.push("remove_control_chars".to_string());
    }

    let before = text.clone();
    text = text.nfc().collect::<String>();
    if text != before {
        ops.push("unicode_nfc".to_string());
    }

    let before = text.clone();
    text = normalize_horizontal_space(&text);
    text = collapse_empty_lines(&text);
    if text != before && !ops.iter().any(|op| op == "normalize_space") {
        ops.push("normalize_space".to_string());
    }

    let trimmed = text.trim().to_string();
    if trimmed != text {
        ops.push("trim".to_string());
        text = trimmed;
    }

    (text, dedupe_ops(ops))
}

fn format_specific_clean(
    file_type: FileType,
    block: &ParsedBlock,
    text: &mut String,
    ops: &mut Vec<String>,
) {
    match file_type {
        FileType::Pdf => clean_pdf_text(block, text, ops),
        FileType::Docx => clean_word_text(block, text, ops),
        FileType::Pptx => clean_ppt_text(block, text, ops),
        FileType::Markdown => clean_markdown_text(block, text, ops),
        FileType::Text => {}
    }
}

fn clean_pdf_text(block: &ParsedBlock, text: &mut String, ops: &mut Vec<String>) {
    if block.block_type == "paragraph" {
        let before = text.clone();
        *text = merge_pdf_line_breaks(text);
        if *text != before {
            ops.push("merge_pdf_line_break".to_string());
        }
    }
    if looks_like_page_number(text) {
        ops.push("remove_pdf_noise".to_string());
    }
}

fn clean_word_text(block: &ParsedBlock, text: &mut String, ops: &mut Vec<String>) {
    if block.block_type == "paragraph" && looks_like_toc_entry(text) {
        ops.push("remove_toc_entry".to_string());
    }
    if block.block_type == "header_footer" {
        ops.push("remove_header_footer".to_string());
    }
}

fn clean_ppt_text(block: &ParsedBlock, text: &mut String, ops: &mut Vec<String>) {
    let lowered = text.to_ascii_lowercase();
    if lowered.contains("click to add") || lowered.contains("单击此处") {
        ops.push("remove_placeholder".to_string());
    }
    if block.block_type == "slide_note" {
        ops.push("preserve_slide_note".to_string());
    }
}

fn clean_markdown_text(block: &ParsedBlock, text: &mut String, ops: &mut Vec<String>) {
    if block.block_type == "html" && is_script_or_style(text) {
        ops.push("remove_script_style".to_string());
    }
    if block.block_type == "comment" || text.trim_start().starts_with("<!--") {
        ops.push("remove_comment".to_string());
    }
    if block.block_type == "image" {
        let alt = markdown_image_alt(text);
        if let Some(alt) = alt {
            *text = alt;
            ops.push("extract_alt_text".to_string());
        }
    }
}

fn removal_reason(
    file_type: FileType,
    block: &ParsedBlock,
    text: &str,
    repeated_pdf_noise: &HashSet<String>,
) -> (bool, Option<String>) {
    let trimmed = text.trim();
    if trimmed.is_empty() {
        return (true, Some("empty_block".to_string()));
    }
    if block.block_type == "header_footer" {
        return (true, Some("header_footer".to_string()));
    }
    if block.block_type == "comment" {
        return (true, Some("comment".to_string()));
    }
    if block.block_type == "html" && is_script_or_style(trimmed) {
        return (true, Some("script_style".to_string()));
    }
    if file_type == FileType::Pdf
        && (looks_like_page_number(trimmed) || repeated_pdf_noise.contains(&noise_key(trimmed)))
    {
        return (true, Some("page_noise".to_string()));
    }
    if file_type == FileType::Docx
        && block.block_type == "paragraph"
        && looks_like_toc_entry(trimmed)
    {
        return (true, Some("toc_entry".to_string()));
    }
    (false, None)
}

fn normalize_horizontal_space(text: &str) -> String {
    text.lines()
        .map(|line| line.split_whitespace().collect::<Vec<_>>().join(" "))
        .collect::<Vec<_>>()
        .join("\n")
}

fn collapse_empty_lines(text: &str) -> String {
    let mut out = String::new();
    let mut empty_count = 0usize;
    for line in text.lines() {
        if line.trim().is_empty() {
            empty_count += 1;
            if empty_count <= 2 {
                out.push('\n');
            }
        } else {
            empty_count = 0;
            if !out.is_empty() && !out.ends_with('\n') {
                out.push('\n');
            }
            out.push_str(line);
        }
    }
    out
}

fn merge_pdf_line_breaks(text: &str) -> String {
    let mut out = String::new();
    for line in text.lines().map(str::trim).filter(|line| !line.is_empty()) {
        if out.ends_with('-') {
            out.pop();
            out.push_str(line);
        } else {
            if !out.is_empty() {
                out.push(' ');
            }
            out.push_str(line);
        }
    }
    out
}

fn looks_like_page_number(text: &str) -> bool {
    let trimmed = text.trim();
    if trimmed.chars().count() > 24 || trimmed.is_empty() {
        return false;
    }
    let lower = trimmed.to_ascii_lowercase();
    let english_page = lower
        .strip_prefix("page ")
        .is_some_and(|tail| tail.trim().chars().all(|c| c.is_ascii_digit()));
    let chinese_page = trimmed
        .strip_prefix('第')
        .and_then(|tail| tail.strip_suffix('页'))
        .is_some_and(|number| {
            !number.is_empty()
                && number.chars().all(|c| {
                    c.is_ascii_digit()
                        || matches!(
                            c,
                            '一' | '二'
                                | '三'
                                | '四'
                                | '五'
                                | '六'
                                | '七'
                                | '八'
                                | '九'
                                | '十'
                                | '百'
                                | '千'
                        )
                })
        });
    english_page || chinese_page
}

fn looks_like_toc_entry(text: &str) -> bool {
    let trimmed = text.trim();
    (trimmed.contains(".....") || trimmed.contains("……")) && trimmed.chars().count() < 200
}

fn repeated_pdf_noise(file_type: FileType, blocks: &[ParsedBlock]) -> HashSet<String> {
    if file_type != FileType::Pdf {
        return HashSet::new();
    }
    let mut pages_by_text = BTreeMap::<String, BTreeSet<i32>>::new();
    for block in blocks {
        let Some(page) = block.page_start else {
            continue;
        };
        let text = block.text.trim();
        if text.is_empty() || text.chars().count() > 120 {
            continue;
        }
        pages_by_text
            .entry(noise_key(text))
            .or_default()
            .insert(page);
    }
    pages_by_text
        .into_iter()
        .filter_map(|(text, pages)| (pages.len() >= 3).then_some(text))
        .collect()
}

fn noise_key(text: &str) -> String {
    text.split_whitespace()
        .collect::<Vec<_>>()
        .join(" ")
        .to_lowercase()
}

fn is_script_or_style(text: &str) -> bool {
    let lower = text.trim_start().to_ascii_lowercase();
    lower.starts_with("<script") || lower.starts_with("<style")
}

fn markdown_image_alt(text: &str) -> Option<String> {
    let start = text.find("![")? + 2;
    let end = text[start..].find(']')? + start;
    let alt = text[start..end].trim();
    (!alt.is_empty()).then(|| alt.to_string())
}

fn dedupe_ops(ops: Vec<String>) -> Vec<String> {
    let mut out = Vec::new();
    for op in ops {
        if !out.contains(&op) {
            out.push(op);
        }
    }
    out
}

File: apps/api-bun/src/document/cleaning.ts
// 移植自 apps/api-rs/src/document/cleaning.rs —— 行为对齐 Rust 原版，错误信息保持一致

// 文本清洗：通用规范化 + 格式特定清理 + 删除判定（页眉页脚/注释/页码噪声/目录等）

import type { FileType, ParsedBlock } from './types.ts';
import { charCount, rustLines, splitWhitespace, trimRust } from './text_utils.ts';

export const CLEANER_VERSION = 'documind-cleaner@0.2.0';

export interface CleanedBlock {
  block: ParsedBlock;
  cleaned_text: string;
  is_removed: boolean;
  remove_reason: string | null;
  cleaning_ops: string[];
}

export interface CleanStats {
  input_blocks: number;
  output_blocks: number;
  removed_blocks: number;
  ops_top: string[];
}

export function cleanBlocks(
  fileType: FileType,
  blocks: ParsedBlock[],
): [CleanedBlock[], CleanStats] {
  const cleaned: CleanedBlock[] = [];
  const opCounts = new Map<string, number>();
  const pdfNoise = repeatedPdfNoise(fileType, blocks);

  for (const block of blocks) {
    const [textBase, opsBase] = commonClean(block.text);
    const text = { value: textBase };
    const ops = { value: opsBase };
    formatSpecificClean(fileType, block, text, ops);

    const [isRemoved, removeReason] = removalReason(fileType, block, text.value, pdfNoise);
    for (const op of ops.value) {
      opCounts.set(op, (opCounts.get(op) ?? 0) + 1);
    }
    cleaned.push({
      block,
      cleaned_text: text.value,
      is_removed: isRemoved,
      remove_reason: removeReason,
      cleaning_ops: ops.value,
    });
  }

  const removedBlocks = cleaned.filter((entry) => entry.is_removed).length;
  const outputBlocks = cleaned.length - removedBlocks;
  const opsTop = [...opCounts.entries()].sort(
    (a, b) => (b[1] - a[1]) || a[0].localeCompare(b[0]),
  );
  return [
    cleaned,
    {
      input_blocks: blocks.length,
      output_blocks: outputBlocks,
      removed_blocks: removedBlocks,
      ops_top: opsTop.slice(0, 8).map(([op]) => op),
    },
  ];
}

export function cleanedBlockMetadata(block: CleanedBlock): Record<string, unknown> {
  return {
    cleaner_version: CLEANER_VERSION,
    cleaning_ops: block.cleaning_ops,
    is_removed: block.is_removed,
    remove_reason: block.remove_reason,
    source_ref: block.block.source_ref,
    source_metadata: block.block.metadata,
  };
}

function commonClean(input: string): [string, string[]] {
  const ops: string[] = [];
  let text = input;

  if (text.startsWith('\uFEFF')) {
    text = text.replace(/^\uFEFF+/, '');
    ops.push('remove_bom');
  }

  if (text.includes('\r\n') || text.includes('\r')) {
    text = text.split('\r\n').join('\n').split('\r').join('\n');
    ops.push('normalize_line_endings');
  }

  const beforeSpace = text;
  text = filterSpecialChars(text);
  if (text !== beforeSpace) {
    ops.push('normalize_space');
    ops.push('remove_control_chars');
  }

  const beforeNfc = text;
  text = text.normalize('NFC');
  if (text !== beforeNfc) {
    ops.push('unicode_nfc');
  }

  const beforeCollapse = text;
  text = collapseEmptyLines(normalizeHorizontalSpace(text));
  if (text !== beforeCollapse && !ops.includes('normalize_space')) {
    ops.push('normalize_space');
  }

  const trimmed = trimRust(text);
  if (trimmed !== text) {
    ops.push('trim');
    text = trimmed;
  }

  return [text, dedupeOps(ops)];
}

/** 对齐 Rust filter_map：NBSP/全角空格 -> 空格；零宽/控制字符删除；tab -> 空格 */
function filterSpecialChars(text: string): string {
  let out = '';
  for (const ch of text) {
    if (ch === '\u00a0' || ch === '\u3000') {
      out += ' ';
    } else if (ch === '\u200b' || ch === '\u200c' || ch === '\u200d' || ch === '\u2060' || ch === '\ufffc') {
      // 删除
    } else if (ch === '\t') {
      out += ' ';
    } else if (ch === '\n') {
      out += '\n';
    } else if (isControlChar(ch)) {
      // 删除
    } else {
      out += ch;
    }
  }
  return out;
}

function isControlChar(ch: string): boolean {
  const codePoint = ch.codePointAt(0)!;
  return (codePoint >= 0x00 && codePoint <= 0x1f) || (codePoint >= 0x7f && codePoint <= 0x9f);
}

function formatSpecificClean(
  fileType: FileType,
  block: ParsedBlock,
  text: { value: string },
  ops: { value: string[] },
): void {
  switch (fileType) {
    case 'pdf':
      cleanPdfText(block, text, ops);
      break;
    case 'docx':
      cleanWordText(block, text, ops);
      break;
    case 'pptx':
      cleanPptText(block, text, ops);
      break;
    case 'md':
      cleanMarkdownText(block, text, ops);
      break;
    case 'txt':
      break;
  }
}

function cleanPdfText(block: ParsedBlock, text: { value: string }, ops: { value: string[] }): void {
  if (block.block_type === 'paragraph') {
    const before = text.value;
    text.value = mergePdfLineBreaks(text.value);
    if (text.value !== before) {
      ops.value.push('merge_pdf_line_break');
    }
  }
  if (looksLikePageNumber(text.value)) {
    ops.value.push('remove_pdf_noise');
  }
}

function cleanWordText(block: ParsedBlock, text: { value: string }, ops: { value: string[] }): void {
  if (block.block_type === 'paragraph' && looksLikeTocEntry(text.value)) {
    ops.value.push('remove_toc_entry');
  }
  if (block.block_type === 'header_footer') {
    ops.value.push('remove_header_footer');
  }
}

function cleanPptText(block: ParsedBlock, text: { value: string }, ops: { value: string[] }): void {
  const lowered = text.value.toLowerCase();
  if (lowered.includes('click to add') || lowered.includes('单击此处')) {
    ops.value.push('remove_placeholder');
  }
  if (block.block_type === 'slide_note') {
    ops.value.push('preserve_slide_note');
  }
}

function cleanMarkdownText(block: ParsedBlock, text: { value: string }, ops: { value: string[] }): void {
  if (block.block_type === 'html' && isScriptOrStyle(text.value)) {
    ops.value.push('remove_script_style');
  }
  if (block.block_type === 'comment' || trimRust(text.value).startsWith('<!--')) {
    ops.value.push('remove_comment');
  }
  if (block.block_type === 'image') {
    const alt = markdownImageAlt(text.value);
    if (alt !== null) {
      text.value = alt;
      ops.value.push('extract_alt_text');
    }
  }
}

function removalReason(
  fileType: FileType,
  block: ParsedBlock,
  text: string,
  repeatedPdfNoise: Set<string>,
): [boolean, string | null] {
  const trimmed = trimRust(text);
  if (trimmed.length === 0) {
    return [true, 'empty_block'];
  }
  if (block.block_type === 'header_footer') {
    return [true, 'header_footer'];
  }
  if (block.block_type === 'comment') {
    return [true, 'comment'];
  }
  if (block.block_type === 'html' && isScriptOrStyle(trimmed)) {
    return [true, 'script_style'];
  }
  if (fileType === 'pdf'
    && (looksLikePageNumber(trimmed) || repeatedPdfNoise.has(noiseKey(trimmed)))) {
    return [true, 'page_noise'];
  }
  if (fileType === 'docx'
    && block.block_type === 'paragraph'
    && looksLikeTocEntry(trimmed)) {
    return [true, 'toc_entry'];
  }
  return [false, null];
}

function normalizeHorizontalSpace(text: string): string {
  return rustLines(text).map((line) => splitWhitespace(line).join(' ')).join('\n');
}

function collapseEmptyLines(text: string): string {
  let out = '';
  let emptyCount = 0;
  for (const line of rustLines(text)) {
    if (trimRust(line).length === 0) {
      emptyCount += 1;
      if (emptyCount <= 2) {
        out += '\n';
      }
    } else {
      emptyCount = 0;
      if (out.length > 0 && !out.endsWith('\n')) {
        out += '\n';
      }
      out += line;
    }
  }
  return out;
}

function mergePdfLineBreaks(text: string): string {
  let out = '';
  for (const line of text.split('\n').map((raw) => trimRust(raw))) {
    if (line.length === 0) continue;
    if (out.endsWith('-')) {
      out = out.slice(0, -1) + line;
    } else {
      if (out.length > 0) {
        out += ' ';
      }
      out += line;
    }
  }
  return out;
}

function looksLikePageNumber(text: string): boolean {
  const trimmed = trimRust(text);
  if (charCount(trimmed) > 24 || trimmed.length === 0) {
    return false;
  }
  const lower = trimmed.toLowerCase();
  const englishPage = lower.startsWith('page ')
    && [...trimRust(lower.slice('page '.length))].every((ch) => ch >= '0' && ch <= '9');
  const chinesePage = (() => {
    if (!trimmed.startsWith('第')) return false;
    const tail = trimmed.slice(1);
    if (!tail.endsWith('页')) return false;
    const number = tail.slice(0, -1);
    if (number.length === 0) return false;
    return [...number].every((ch) => (ch >= '0' && ch <= '9') || '一二三四五六七八九十百千'.includes(ch));
  })();
  return englishPage || chinesePage;
}

function looksLikeTocEntry(text: string): boolean {
  const trimmed = trimRust(text);
  return (trimmed.includes('.....') || trimmed.includes('……')) && charCount(trimmed) < 200;
}

function repeatedPdfNoise(fileType: FileType, blocks: ParsedBlock[]): Set<string> {
  if (fileType !== 'pdf') return new Set();
  const pagesByText = new Map<string, Set<number>>();
  for (const block of blocks) {
    if (block.page_start === null) continue;
    const text = trimRust(block.text);
    if (text.length === 0 || charCount(text) > 120) continue;
    const key = noiseKey(text);
    const pages = pagesByText.get(key);
    if (pages) {
      pages.add(block.page_start);
    } else {
      pagesByText.set(key, new Set([block.page_start]));
    }
  }
  const out = new Set<string>();
  for (const [text, pages] of pagesByText) {
    if (pages.size >= 3) out.add(text);
  }
  return out;
}

function noiseKey(text: string): string {
  return splitWhitespace(text).join(' ').toLowerCase();
}

function isScriptOrStyle(text: string): boolean {
  const lower = trimRust(text).toLowerCase();
  return lower.startsWith('<script') || lower.startsWith('<style');
}

function markdownImageAlt(text: string): string | null {
  const start = text.indexOf('![');
  if (start < 0) return null;
  const end = text.indexOf(']', start + 2);
  if (end < 0) return null;
  const alt = trimRust(text.slice(start + 2, end));
  return alt.length > 0 ? alt : null;
}

function dedupeOps(ops: string[]): string[] {
  const out: string[] = [];
  for (const op of ops) {
    if (!out.includes(op)) out.push(op);
  }
  return out;
}
